sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala

```diff
@@ -662,6 +665,154 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
+
+          // batch 2: check that things work well when the sink log gets compacted
+          q1AddData("keep4"),
+          Assert {
+            // compact interval is 3, so file "2.compact" should exist
+            new File(q1_outputDir, s"${FileStreamSink.metadataDir}/2.compact").exists()
+          },
+          q2ProcessAllAvailable(),
+          CheckAnswer("keep2", "keep3", "keep4"),
+
+          // stop q1 manually
+          Execute { _ => q1.stop() }
+        )
+      }
+    }
+  }
+
+  test("read partitioned data from outputs of another streaming query") {
+    withTempDirs { case (dir, tmp) =>
+      // q1 is a streaming query that reads from memory and writes to partitioned json files
+      val q1_source = MemoryStream[(String, String)]
+      val q1_checkpointDir = new File(dir, "q1_checkpointDir").getCanonicalPath
+      val q1_outputDir = new File(dir, "q1_outputDir").getCanonicalPath
+      val q1 =
+        q1_source
+          .toDF()
+          .select($"_1" as "partition", $"_2" as "value")
+          .writeStream
+          .option("checkpointLocation", q1_checkpointDir)
+          .partitionBy("partition")
+          .format("json")
+          .start(q1_outputDir)
+
+      // q2 is a streaming query that reads q1's partitioned json outputs
+      val schema = new StructType().add("value", StringType).add("partition", StringType)
+      val q2 = createFileStream("json", q1_outputDir, Some(schema)).filter($"value" contains "keep")
+
+      def q1AddData(data: (String, String)*): StreamAction =
+        Execute { _ =>
+          q1_source.addData(data)
+          q1.processAllAvailable()
+        }
+      def q2ProcessAllAvailable(): StreamAction = Execute { q2 => q2.processAllAvailable() }
+
+      testStream(q2)(
+        // batch 0: append to a new partition=foo, should read value and partition
+        q1AddData(("foo", "drop1"), ("foo", "keep2")),
+        q2ProcessAllAvailable(),
+        CheckAnswer(("keep2", "foo")),
+
+        // batch 1: append to same partition=foo, should read value and partition
+        q1AddData(("foo", "keep3")),
+        q2ProcessAllAvailable(),
+        CheckAnswer(("keep2", "foo"), ("keep3", "foo")),
+
+        // batch 2: append to a different partition=bar, should read value and partition
+        q1AddData(("bar", "keep4")),
+        q2ProcessAllAvailable(),
+        CheckAnswer(("keep2", "foo"), ("keep3", "foo"), ("keep4", "bar")),
+
+        // stop q1 manually
+        Execute { _ => q1.stop() }
+      )
+    }
+  }
+
+  test("start before another streaming query, and read its output") {
+    withTempDirs { case (dir, tmp) =>
+      // q1 is a streaming query that reads from memory and writes to text files
+      val q1_source = MemoryStream[String]
+      val q1_checkpointDir = new File(dir, "q1_checkpointDir").getCanonicalPath
+      val q1_outputDir = new File(dir, "q1_outputDir")
+      assert(q1_outputDir.mkdir())                    // prepare the output dir for q2 to read
```

Comment:
nit: just put the command following the statement with 1 space. Using the current format is hard to maintain in future because it requires to align comments. Same for other comments.